### PR TITLE
Ripple

### PR DESCRIPTION
**Pull request by carlcherry**
Added Ripple support in the core C library along with unit tests.

### CODE REVIEW

**Comment by EBGToo** [comment]
I confirmed that a merge to 'crypto' builds cleanly and that all unit tests, included newly added 'ripple' tests complete successfully.

Majority of work is in a 'ripple/' directory.  A couple of functions are added to BRBase58.[ch] - to generalize encode/decode to use the 'ripple alphabet'.

File: ripple/BRRippleAccount.c [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    BRBIP39DeriveKey(seed.u8, paperKey, NULL); // no passphrase
+    return seed;
+}
+
+extern BRKey deriveRippleKeyFromSeed (UInt512 seed, uint32_t index)
+{
+    BRKey privateKey;
+    
+    // The BIP32 privateKey for m/44'/60'/0'/0/index
+    BRBIP32PrivKeyPath(&privateKey, &seed, sizeof(UInt512), 5,
```
Comment (michael-brd) [inline ripple/BRRippleAccount.c]: Would having definitions for these values make sense?

File: ripple/BRRippleAccount.c [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+    // 3. Use the first 4 bytes of the hash as checksum and append to the bytes
+    uint8_t input[25];
+    input[0] = 0; // Ripple address type
+    memcpy(&input[1], address.bytes, 20);
+    uint8_t hash[32];
+    BRSHA256_2(hash, input, 21);
+    memcpy(&input[21], hash, 4);
+
+    // Now base58 encode the result
+    rippleEncodeBase58(string, 35, input, 25);
```
Comment (michael-brd) [inline ripple/BRRippleAccount.c]: Would having a define for the address string length make sense? Not sure if it used elsewhere.

File: ripple/BRRippleAccount.c [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+}
+
+/**
+ * Optional way to create the BRKey
+ *
+ * 1. The normal way using the mnemonic paper_key
+ * 2. In DEBUG mode unit tests can pass in a private key instead
+ *
+ */
+BRKey getKey(const char* paperKey)
```
Comment (michael-brd) [inline ripple/BRRippleAccount.c]: Would splitting this into two functions (with one only exposed in debug builds) make sense from a testing point of view?

File: ripple/BRRippleAccount.c [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+}
+
+extern BRRippleAddress rippleAccountGetAddress(BRRippleAccount account)
+{
+    BRRippleAddress address;
+    memcpy(address.bytes, account->raw.bytes, sizeof(address.bytes));
+    return(address);
+}
+
+extern int rippleAccountGetAddressString(BRRippleAccount account, char * rippleAddress, int length)
```
Comment (michael-brd) [inline ripple/BRRippleAccount.c]: Would using size_t make sense here?

File: ripple/BRRippleAccount.c [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    memset(address.bytes, 0x00, sizeof(address.bytes));
+    // Work backwards from this ripple address (string) to what is
+    // known as the acount ID (20 bytes)
+    rippleAddressStringToAddress(rippleAddressString, &address);
+    return address;
+}
+
+extern int // 1 if equal
+rippleAddressEqual (BRRippleAddress a1, BRRippleAddress a2) {
+    return 0 == memcmp (a1.bytes, a2.bytes, 20);
```
Comment (michael-brd) [inline ripple/BRRippleAccount.c]: Would sizeof make sense here?

**Pull request merged**